// @tanstack/router-core/dist/cjs/index.cjs
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const require_utils = require("./utils.cjs");
const require_invariant = require("./invariant.cjs");
const require_path = require("./path.cjs");
const require_not_found = require("./not-found.cjs");
const require_scroll_restoration = require("./scroll-restoration.cjs");
const require_qss = require("./qss.cjs");
const require_searchParams = require("./searchParams.cjs");
const require_root = require("./root.cjs");
const require_redirect = require("./redirect.cjs");
const require_rewrite = require("./rewrite.cjs");
const require_stores = require("./stores.cjs");
const require_router = require("./router.cjs");
const require_defer = require("./defer.cjs");
const require_link = require("./link.cjs");
const require_manifest = require("./manifest.cjs");
const require_Matches = require("./Matches.cjs");
const require_route = require("./route.cjs");
const require_config = require("./config.cjs");
const require_searchMiddleware = require("./searchMiddleware.cjs");
const require_hash_scroll = require("./hash-scroll.cjs");
const require_transformer = require("./ssr/serializer/transformer.cjs");
const require_RawStream = require("./ssr/serializer/RawStream.cjs");
const require_seroval_plugins = require("./ssr/serializer/seroval-plugins.cjs");
exports.BaseRootRoute = require_route.BaseRootRoute;
exports.BaseRoute = require_route.BaseRoute;
exports.BaseRouteApi = require_route.BaseRouteApi;
exports.DEFAULT_PROTOCOL_ALLOWLIST = require_utils.DEFAULT_PROTOCOL_ALLOWLIST;
exports.PathParamError = require_router.PathParamError;
exports.RawStream = require_RawStream.RawStream;
exports.RouterCore = require_router.RouterCore;
exports.SearchParamError = require_router.SearchParamError;
exports.TSR_DEFERRED_PROMISE = require_defer.TSR_DEFERRED_PROMISE;
exports.buildDevStylesUrl = require_utils.buildDevStylesUrl;
exports.cleanPath = require_path.cleanPath;
exports.composeRewrites = require_rewrite.composeRewrites;
exports.createControlledPromise = require_utils.createControlledPromise;
exports.createNonReactiveMutableStore = require_stores.createNonReactiveMutableStore;
exports.createNonReactiveReadonlyStore = require_stores.createNonReactiveReadonlyStore;
exports.createRawStreamDeserializePlugin = require_RawStream.createRawStreamDeserializePlugin;
exports.createRawStreamRPCPlugin = require_RawStream.createRawStreamRPCPlugin;
exports.createRouterConfig = require_config.createRouterConfig;
exports.createSerializationAdapter = require_transformer.createSerializationAdapter;
exports.decode = require_qss.decode;
exports.deepEqual = require_utils.deepEqual;
exports.defaultGetScrollRestorationKey = require_scroll_restoration.defaultGetScrollRestorationKey;
exports.defaultParseSearch = require_searchParams.defaultParseSearch;
exports.defaultSerializeError = require_router.defaultSerializeError;
exports.defaultSerovalPlugins = require_seroval_plugins.defaultSerovalPlugins;
exports.defaultStringifySearch = require_searchParams.defaultStringifySearch;
exports.defer = require_defer.defer;
exports.encode = require_qss.encode;
exports.escapeHtml = require_utils.escapeHtml;
exports.exactPathTest = require_path.exactPathTest;
exports.executeRewriteInput = require_rewrite.executeRewriteInput;
exports.functionalUpdate = require_utils.functionalUpdate;
exports.getAssetCrossOrigin = require_manifest.getAssetCrossOrigin;
exports.getElementScrollRestorationEntry = require_scroll_restoration.getElementScrollRestorationEntry;
exports.getInitialRouterState = require_router.getInitialRouterState;
exports.getLocationChangeInfo = require_router.getLocationChangeInfo;
exports.getMatchedRoutes = require_router.getMatchedRoutes;
exports.handleHashScroll = require_hash_scroll.handleHashScroll;
exports.interpolatePath = require_path.interpolatePath;
exports.invariant = require_invariant.invariant;
exports.isDangerousProtocol = require_utils.isDangerousProtocol;
exports.isMatch = require_Matches.isMatch;
exports.isModuleNotFoundError = require_utils.isModuleNotFoundError;
exports.isNotFound = require_not_found.isNotFound;
exports.isPlainArray = require_utils.isPlainArray;
exports.isPlainObject = require_utils.isPlainObject;
exports.isRedirect = require_redirect.isRedirect;
exports.isResolvedRedirect = require_redirect.isResolvedRedirect;
exports.joinPaths = require_path.joinPaths;
exports.lazyFn = require_router.lazyFn;
exports.makeSerovalPlugin = require_transformer.makeSerovalPlugin;
exports.makeSsrSerovalPlugin = require_transformer.makeSsrSerovalPlugin;
exports.notFound = require_not_found.notFound;
exports.parseRedirect = require_redirect.parseRedirect;
exports.parseSearchWith = require_searchParams.parseSearchWith;
exports.preloadWarning = require_link.preloadWarning;
exports.redirect = require_redirect.redirect;
exports.removeTrailingSlash = require_path.removeTrailingSlash;
exports.replaceEqualDeep = require_utils.replaceEqualDeep;
exports.resolveManifestAssetLink = require_manifest.resolveManifestAssetLink;
exports.resolvePath = require_path.resolvePath;
exports.retainSearchParams = require_searchMiddleware.retainSearchParams;
exports.rootRouteId = require_root.rootRouteId;
exports.scrollRestorationCache = require_scroll_restoration.scrollRestorationCache;
exports.setupScrollRestoration = require_scroll_restoration.setupScrollRestoration;
exports.storageKey = require_scroll_restoration.storageKey;
exports.stringifySearchWith = require_searchParams.stringifySearchWith;
exports.stripSearchParams = require_searchMiddleware.stripSearchParams;
exports.trailingSlashOptions = require_router.trailingSlashOptions;
exports.trimPath = require_path.trimPath;
exports.trimPathLeft = require_path.trimPathLeft;
exports.trimPathRight = require_path.trimPathRight;
