// @tanstack/router-core/dist/cjs/ssr/server.cjs
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const require_ssr_server = require("./ssr-server.cjs");
const require_createRequestHandler = require("./createRequestHandler.cjs");
const require_handlerCallback = require("./handlerCallback.cjs");
const require_transformStreamWithRouter = require("./transformStreamWithRouter.cjs");
exports.attachRouterServerSsrUtils = require_ssr_server.attachRouterServerSsrUtils;
exports.createRequestHandler = require_createRequestHandler.createRequestHandler;
exports.defineHandlerCallback = require_handlerCallback.defineHandlerCallback;
exports.getNormalizedURL = require_ssr_server.getNormalizedURL;
exports.getOrigin = require_ssr_server.getOrigin;
exports.transformPipeableStreamWithRouter = require_transformStreamWithRouter.transformPipeableStreamWithRouter;
exports.transformReadableStreamWithRouter = require_transformStreamWithRouter.transformReadableStreamWithRouter;
exports.transformStreamWithRouter = require_transformStreamWithRouter.transformStreamWithRouter;
