// @kitware/vtk.js/Widgets/Representations/ScaleTransformHandleRepresentation.js
import { m as macro } from '../../macros2.js';
import vtkGlyphRepresentation from './GlyphRepresentation.js';
import vtkCubeSource from '../../Filters/Sources/CubeSource.js';
import vtkTransformHandleSource from './TranslateTransformHandleRepresentation/TransformHandleSource.js';

// ----------------------------------------------------------------------------
// vtkScaleTransformHandleRepresentation methods
// ----------------------------------------------------------------------------

function vtkScaleTransformHandleRepresentation(publicAPI, model) {
  // Set our className
  model.classHierarchy.push('vtkScaleTransformHandleRepresentation');
}

// ----------------------------------------------------------------------------
// Object factory
// ----------------------------------------------------------------------------

// ----------------------------------------------------------------------------
function defaultValues(initialValues) {
  const source = vtkTransformHandleSource.newInstance({
    height: initialValues.height ?? 1,
    radius: initialValues.radius ?? 1,
    resolution: initialValues.glyphResolution ?? 12,
    direction: [0, 0, 1]
  });
  const cube1 = vtkCubeSource.newInstance(initialValues.cubeSource);
  const cube2 = vtkCubeSource.newInstance(initialValues.cubeSource);
  source.addInputConnection(cube1.getOutputPort());
  source.addInputConnection(cube2.getOutputPort());
  return {
    _pipeline: {
      glyph: source
    },
    ...initialValues
  };
}
function extend(publicAPI, model, initialValues = {}) {
  vtkGlyphRepresentation.extend(publicAPI, model, defaultValues(initialValues));

  // Object specific methods
  vtkScaleTransformHandleRepresentation(publicAPI, model);
}

// ----------------------------------------------------------------------------

const newInstance = macro.newInstance(extend, 'vtkScaleTransformHandleRepresentation');

// ----------------------------------------------------------------------------

var vtkScaleTransformHandleRepresentation$1 = {
  newInstance,
  extend
};

export { vtkScaleTransformHandleRepresentation$1 as default, extend, newInstance };
