// @kitware/vtk.js/Widgets/Representations/TranslateTransformHandleRepresentation.js
import { m as macro } from '../../macros2.js';
import vtkGlyphRepresentation from './GlyphRepresentation.js';
import vtkConeSource from '../../Filters/Sources/ConeSource.js';
import vtkTransformHandleSource from './TranslateTransformHandleRepresentation/TransformHandleSource.js';

// ----------------------------------------------------------------------------
// vtkTranslateTransformHandleRepresentation methods
// ----------------------------------------------------------------------------

function vtkTranslateTransformHandleRepresentation(publicAPI, model) {
  // Set our className
  model.classHierarchy.push('vtkTranslateTransformHandleRepresentation');
}

// ----------------------------------------------------------------------------
// Object factory
// ----------------------------------------------------------------------------

// ----------------------------------------------------------------------------
function defaultValues(initialValues) {
  const source = vtkTransformHandleSource.newInstance({
    height: initialValues.height ?? 1,
    radius: initialValues.radius ?? 1,
    resolution: initialValues.glyphResolution ?? 12,
    direction: [0, 0, 1]
  });
  const cone1 = vtkConeSource.newInstance(initialValues.coneSource);
  const cone2 = vtkConeSource.newInstance(initialValues.coneSource);
  source.addInputConnection(cone1.getOutputPort());
  source.addInputConnection(cone2.getOutputPort());
  return {
    _pipeline: {
      glyph: source
    },
    ...initialValues
  };
}
function extend(publicAPI, model, initialValues = {}) {
  vtkGlyphRepresentation.extend(publicAPI, model, defaultValues(initialValues));

  // Object specific methods
  vtkTranslateTransformHandleRepresentation(publicAPI, model);
}

// ----------------------------------------------------------------------------

const newInstance = macro.newInstance(extend, 'vtkTranslateTransformHandleRepresentation');

// ----------------------------------------------------------------------------

var vtkTranslateTransformHandleRepresentation$1 = {
  newInstance,
  extend
};

export { vtkTranslateTransformHandleRepresentation$1 as default, extend, newInstance };
